workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityCommandSenderReceiverGenerator.tt

```diff
@@ -286,6 +287,20 @@ namespace <#= qualifiedNamespace #>
                 callbackSystem.RegisterCommandResponseCallback(requestId, wrappedCallback);
             }
         }
+
+        public async Task<<#= receivedCommandResponseType #>> Send<#= commandDetails.CommandName #>Command(<#= commandRequest #> request)
+        {
+            var requestId = commandSender.SendCommand(request, entity);
+            <#= receivedCommandResponseType #> response = default(<#= receivedCommandResponseType #>);
+            await Task.Run(() =>
+            {
+                while (!commandSender.TryGetReceivedResponseForRequestId(requestId, out response))
+                {
```

Comment:
we might want to add the option to add a `Task.Delay(..)` inbetween so that we are not polling all the time

Reply:
This busy waiting is definitely something to avoid.
Why does the `commandSender` not also have an `async` method to await the response (keep pushing async down as far as it makes sense.)
You could have an `AutoResetEvent` that is signaled when the command response op comes through and is stored in your table - http://dotnetpattern.com/threading-autoresetevent

Reply:
completely rewritten it after some more discussion. Now using `TaskCompletionSource` to get rid of `Task.Run` and the while loop